src/v/cloud_storage/remote_segment_index.h

```diff
@@ -0,0 +1,1139 @@
+ * compiler since the loops that process the single row can be
+ * easily unrolled or vectorized since the number of iterations is
+ * known at compile time. Most bit-packing required by FoR can be
+ * done as a series of simple operations (shifts/loads/stores) in a
+ * loop that doesn't require any branching.
+ *
+ * Also, because bit-packing routines work with rows of 16 elements
+ * they always start and stop on a byte boundary. This simplifies the
+ * code quite a lot.
+ *
+ * The FoR encoding memory layout is not conventional. It's not
+ * placing values one after another. It uses the following schema
+ * instead. Consider the following size classes: 64, 32, 16, and
+ * 8-bits. Each 64-bit value can be represented as a series of those + some
+ * reminder. For instance, 18-bit value can be represented as 16-bit
+ * value + 2-bit reminder, 47-bit value can be represented as 32-bit
+ * value + 8-bit value + 7-bit reminder, etc.
+ *
+ * The bit-packing algorithm works in the following way. First, it
+ * calculates minimal number of bits that can be used to store any
+ * value in a row. Then the number of bits is factored into one or
+ * several size classes + reminder. After that the algorithm writes
+ * the number of bits that corresponds to the largest size class, then
+ * the next one, etc. After that it writes the reminder of every element.
+ * This means that the data from all 16 values is interliving. The number
+ * of bits used to represent the row is stored using 8 bits.
+ *
+ * Example: let's say that we have a row that requires us to use 59
+ * bits per element. Bit-packing will go like this:
+ * - write first 32-bits from every value in a row (64 bytes total);
+ * - write next 16-bits from every value in a row (32 bytes total);
+ * - write next 8-bits from every value in a row (16 bytes total);
+ * - write the remaining 3-bits from every value (6 bytes total);
+ * The result is represented using 118 bytes.
+ *
+ * One advantage of this approach is that it requires less code to
+ * implement. It needs bit-packing functions that can pack 1-7 bits
+ * and also 8, 16, 32, and 64 bits. All possible bit-packing arrangements
+ * from 0 to 64 bits can be produced using this functions.
+ * The alternative to this is to implement 63 bit-packing functions that
+ * can pack all possible values. The approach used here requires only
+ * 7 custom bit-packing functions + 4 bit-packing functions for different
+ * size classes (8, 16, 32, 64) which is much easier to impelment and
+ * test.
+ *
+ * It's also beneficial for further compression using general purpose
+ * compression algorithms (e.g. zstd or lz4). If the values have some
+ * common substructure (e.g. the lowest bits are zeroed) this common bits
+ * will be stored together and the compression alg. could take advantage
+ * of that.
+ *
+ * The compressed data is stored internally using an iobuf. This iobuf
+ * can be copied and pushed to the decoder to decompress the values.
```

Comment:
>  * The compression ratio depends on data and for redpanda offsets it's expected to be around 10% to 15%.

are all batches in the segment represented in the index? if so, do you think there would be value in using (if you aren't already) indexing at a coarse granularity similar to what we do in the segment index? for example, the segment index only records batches that are at least 32KB offset away from each other. Then we always can seek close to the target offset, but also keep the index small for small batch workloads.

Reply:
ahh cool, I missed that.